src/selectors/modular-pipelines.js

```diff
@@ -1,34 +1,220 @@
+      .length,
+  })
+);
+
+/**
+ * List all the child nodes in each modular pipeline
+ */
+export const getModularPipelineChildren = createSelector(
+  [getModularPipelineIDs, getNodeIDs, getNodeModularPipelines],
+  (modularPipelineIDs, nodeIDs, nodeModularPipelines) => {
+    const modPipNodes = arrayToObject(modularPipelineIDs, () => ({}));
+    nodeIDs.forEach((nodeID) => {
+      nodeModularPipelines[nodeID]?.forEach((modPipID) => {
+        if (!modPipNodes[modPipID]) {
+          modPipNodes[modPipID] = {};
+        }
+        modPipNodes[modPipID][nodeID] = true;
+      });
+    });
+    return modPipNodes;
+  }
+);
+
+/**
+ * List all the parent modular pipelines for a given modular pipeline
+ */
+export const getModularPipelineParents = createSelector(
+  [getModularPipelineIDs],
+  (modularPipelineIDs) =>
+    arrayToObject(modularPipelineIDs, (modPipID) =>
+      modPipID.split('.').reduce((parents, part, i, parts) => {
+        if (i < parts.length - 1) {
+          let parent = part;
+          if (i > 0) {
+            parent = parents[parents.length - 1] + '.' + parent;
+          }
+          parents.push(parent);
+        }
+        return parents;
+      }, [])
+    )
+);
+
+/**
+ * Set disabled status if the node is specifically hidden, and/or via a tag/view/type/modularPipeline
+ */
+export const getModularPipelineParentsContracted = createSelector(
+  [
+    getModularPipelineIDs,
+    getModularPipelineParents,
+    getModularPipelineContracted,
+  ],
+  (modularPipelineIDs, modularPipelineParents, modularPipelineContracted) =>
+    arrayToObject(modularPipelineIDs, (modPipID) =>
+      modularPipelineParents[modPipID].some(
+        (id) => modularPipelineContracted[id]
+      )
+    )
+);
+
+/**
+ * Get a list of input/output edges of visible modular pipeline pseudo-nodes
+ * by examining the edges of their childen, and also create transitive edges
+ * between collapsed modular pipelines
+ */
+export const getModularPipelineEdges = createSelector(
+  [
+    getModularPipelineIDs,
+    getModularPipelineChildren,
+    getNodeModularPipelines,
+    getEdgeIDs,
+    getEdgeSources,
+    getEdgeTargets,
+  ],
+  (
+    modularPipelineIDs,
+    modularPipelineChildren,
+    nodeModularPipelines,
+    edgeIDs,
+    edgeSources,
+    edgeTargets
+  ) => {
+    // List of new edges generated from modular pipelines:
+    const edges = {
+      ids: {},
+      sources: {},
+      targets: {},
+    };
+
+    const addNewEdge = (source, target) => {
+      const id = [source, target].join('|');
+      edges.ids[id] = true;
+      edges.sources[id] = source;
+      edges.targets[id] = target;
+    };
+    modularPipelineIDs.forEach((modPipID) => {
+      const modPipNodes = modularPipelineChildren[modPipID];
+      edgeIDs.forEach((edgeID) => {
+        const source = edgeSources[edgeID];
+        const target = edgeTargets[edgeID];
+        if (modPipNodes[target] && !modPipNodes[source]) {
```

Comment:
The comment below each on was the comment explaining this (I never know where to put the comment with `if` statements), but it evidently wasn't clear so I've made these lines into named variables so that comments are unnecessary.